src/osx/core/colour.cpp

```diff
@@ -13,127 +13,244 @@
+        m_red = components[0];
+        m_green = components[1];
+        m_blue = components[2];
+
+        if (noComp == 4)
+            m_alpha = components[3];
+    }
+}
+
+#define M_COLDATA static_cast<wxColorRefData*>(m_refData)
+
 #if wxOSX_USE_COCOA_OR_CARBON
 wxColour::wxColour(const RGBColor& col)
 {
-    InitRGBColor(col);
+    CGFloat components[4] = { (CGFloat)(col.red / 65535.0), (CGFloat)(col.green / 65535.0),
+        (CGFloat)(col.blue / 65535.0), (CGFloat)1.0 };
+    m_refData = new wxCGColorRefData(components);
 }
 #endif
 
 wxColour::wxColour(CGColorRef col)
 {
-    InitCGColorRef(col);
+    m_refData = new wxCGColorRefData(col);
 }
 
-#if wxOSX_USE_COCOA_OR_CARBON
-void wxColour::GetRGBColor( RGBColor *col ) const
+wxColour::ChannelType wxColour::Red() const
 {
-    col->red = (m_red << 8) + m_red;
-    col->blue = (m_blue << 8) + m_blue;
-    col->green = (m_green << 8) + m_green;
+    return wxRound(M_COLDATA->Red() * 255.0);
 }
 
-wxColour& wxColour::operator=(const RGBColor& col)
+wxColour::ChannelType wxColour::Green() const
 {
-    InitRGBColor(col);
-    return *this;
+    return wxRound(M_COLDATA->Red() * 255.0);
 }
-#endif
 
-wxColour& wxColour::operator=(CGColorRef col)
+wxColour::ChannelType wxColour::Blue() const
 {
-    InitCGColorRef(col);
-    return *this;
+    return wxRound(M_COLDATA->Red() * 255.0);
 }
 
-wxColour& wxColour::operator=(const wxColour& col)
+wxColour::ChannelType wxColour::Alpha() const
 {
-    m_red = col.m_red;
-    m_green = col.m_green;
-    m_blue = col.m_blue;
-    m_alpha = col.m_alpha;
-    m_cgColour = col.m_cgColour;
-    return *this;
+    return wxRound(M_COLDATA->Red() * 255.0);
 }
 
-void wxColour::InitRGBA (ChannelType r, ChannelType g, ChannelType b, ChannelType a)
+#if wxOSX_USE_COCOA_OR_CARBON
+void wxColour::GetRGBColor(RGBColor* col) const
 {
-    m_red = r;
-    m_green = g;
-    m_blue = b;
-    m_alpha = a ;
-    
-    CGColorRef col = 0 ;
-    CGFloat components[4] = { (CGFloat)(r / 255.0), (CGFloat) (g / 255.0), (CGFloat) (b / 255.0), (CGFloat) (a / 255.0) } ;
-    col = CGColorCreate( wxMacGetGenericRGBColorSpace() , components ) ;
-    
-    wxASSERT_MSG(col != NULL, "Invalid CoreGraphics Color");
-    m_cgColour.reset( col );
+    col->red = M_COLDATA->Red() * 65535.0;
+    col->blue = M_COLDATA->Blue() * 65535.0;
+    col->green = M_COLDATA->Green() * 65535.0;
 }
+#endif
 
-#if wxOSX_USE_COCOA_OR_CARBON
-void wxColour::InitRGBColor( const RGBColor& col )
+bool wxColour::IsOk() const
 {
-    m_red = col.red >> 8;
-    m_blue = col.blue >> 8;
-    m_green = col.green >> 8;
-    m_alpha = wxALPHA_OPAQUE;
-    CGColorRef cfcol;
-    CGFloat components[4] = { (CGFloat)(col.red / 65535.0), (CGFloat)(col.green / 65535.0),
-        (CGFloat)(col.blue / 65535.0), (CGFloat) 1.0 } ;
-    cfcol = CGColorCreate( wxMacGetGenericRGBColorSpace() , components ) ;
+    if (m_refData)
+        return M_COLDATA->IsOk();
```

Comment:
What's the purpose of having a ref data which is "not ok"? I'd rather ensure that it's always ok, deleting it if it isn't.

Reply:
OK, I'll move the check to initialization and delete if no native color is created